cameo/strain_design/heuristic/evolutionary/optimization.py

```diff
@@ -806,3 +845,223 @@ def __init__(self, reaction_db=None, *args, **kwargs):
+    co-factor. This class implements a search for reactions when swapped improve the given objective. Briefly,
+    the approach is to
+
+    - find reactions that can safely be swapped e.g. (nad_c -> nadp_c, nadh_c -> nadph_c) without resulting in an
+      infeasible model
+
+    - add reactions that have the co-factors swapped and then by a search algorithm switching one off in favor of the
+      other
+
+   The implementation here differs from that in [1]_ in that we use a general purpose search algorithm rather than
+   formulating the search as a mixed integer linear programming problem.
+
+   References
+   ----------
+   .. [1] King, Zachary A., and Adam M. Feist. “Optimizing Cofactor Specificity of Oxidoreductase Enzymes for the
+      Generation of Microbial Production Strains—OptSwap.” Industrial Biotechnology 9, no. 4 (August 1,
+      2013): 236–46. ---doi:10.1089/ind.2013.0005.
+
+    Parameters
+    ----------
+    model : SolverBasedModel
+       the model to operator on
+    cofactor_id_swaps : list
+       a list of length 2 where the first element contains a list of metabolite identifiers to swap-out by replacing
+       them with the metabolites listed in the second element of the same list, see e.g. `NADH_NADPH` which is also
+       the default
+    candidate_reactions : list
+       reactions to consider for co-factor swap - if not given then search for all reactions that include the given
+       cofactors
+    skip_reactions : list
+       reactions to not consider for co-factor swap, defaults to the objective function if not provided
+    args, kwargs : keyword arguments
+       passed on to super-classes, see in particular `objective_function`, `heuristic_method`, `termination`,
+       `simulation_method`, `wt_reference`, of `HeuristicOptimization` and `max_size` of `HeuristicOptimization.run`
+
+    Examples
+    --------
+    >>> from cameo import models
+    >>> model = models.bigg.iJO1366
+    >>> from cameo.strain_design.heuristic.evolutionary.objective_functions import biomass_product_coupled_yield
+    >>> py = biomass_product_coupled_yield(model.reactions.EX_cys__L_e, model.reactions.EX_glc__D_e)
+    >>> swap_optimization = CofactorSwapOptimization(model=model, objective_function=py)
+    >>> swap_optimization.run(max_evaluations=2000, max_size=3)
+    """
+
+    def __init__(self, model, cofactor_id_swaps=NADH_NADPH, candidate_reactions=None,
+                 skip_reactions=None, *args, **kwargs):
+        super(self.__class__, self).__init__(model=model, knockout_evaluator=SwapEvaluator, *args, **kwargs)
+        self.model = SwapperModel(model, cofactor_id_swaps, candidate_reactions, skip_reactions)
+        self._ko_type = SWAP_KNOCKOUT_TYPE
+        self.representation = list(self.model.swapped_reactions.keys())
+        self._decoder = decoders.ReactionKnockoutDecoder(self.representation, self.model)
+
+
+class SwapperModel(SolverBasedModel):
```

Comment:
I understand it might seem unnecessary and there are for sure many ways to do this - however, swapping a reaction as a method of a model seemed oop and convenient to me as it manages all the arguments in a straight-forward fashion, avoiding reliance on `*args, **kwargs` and ilk - I did not notice significant draw-backs with subclassing the model - did you?